File: src/tools/pedestrianTool.ts
import type { Point } from '../sim/geometry';
import { EMPTY_PREVIEW, cursorUrl, type PointerInfo, type Tool, type ToolContext, type ToolPreview } from './types';

/**
 * Adds pedestrians, ported from controller/PedestrianMouseListener: a click drops
 * an n x n block and dragging paints continuously, so a crowd can be laid down in
 * one gesture.
 */
export class PedestrianTool implements Tool {
  readonly id = 'pedestrian' as const;
  readonly cursor = cursorUrl('ped', 16, 16);
  private painting = false;
  /** Ghost dots for the block under the cursor, as drawTemporaryPedestrians did. */
  private ghost: Point[] = [];

  onPointerDown(e: PointerInfo, ctx: ToolContext): void {
    if (e.buttons !== 1) return;
    this.painting = true;
    ctx.addPedestrians(e.world);
    this.ghost = ctx.pedestrianBlock(e.world);
  }

  onPointerMove(e: PointerInfo, ctx: ToolContext): void {
    if (this.painting && e.buttons !== 0) ctx.addPedestrians(e.world);
    // Recomputed after placing, so the preview shows only spots still free.
    this.ghost = ctx.pedestrianBlock(e.world);
    ctx.requestRender();
  }

  onPointerUp(): void { this.painting = false; }

  cancel(): void {
    this.painting = false;
    this.ghost = [];
  }

  preview(): ToolPreview {
    return { ...EMPTY_PREVIEW, pendingPedestrians: this.ghost };
  }
}

File: ios/Sources/WalkyCore/Tools/PedestrianTool.swift
import Foundation

/// Adds pedestrians, from `controller/PedestrianMouseListener`: a tap drops an
/// n x n block and dragging paints continuously, so a crowd can be laid down in
/// one gesture. Ports `src/tools/pedestrianTool.ts`.
@MainActor
public final class PedestrianTool: Tool {
  public let id = ToolId.pedestrian
  private var painting = false
  /// Ghost dots for the block under the cursor, as `drawTemporaryPedestrians` did.
  private var ghost: [Point] = []

  public init() {}

  public func onPointerDown(_ e: PointerInfo, _ ctx: ToolContext) {
    if e.buttons != 1 { return }
    painting = true
    ctx.addPedestrians(e.world)
    ghost = ctx.pedestrianBlock(e.world, nil)
  }

  public func onPointerMove(_ e: PointerInfo, _ ctx: ToolContext) {
    if painting && e.buttons != 0 { ctx.addPedestrians(e.world) }
    // Recomputed after placing, so the preview shows only spots still free.
    ghost = ctx.pedestrianBlock(e.world, nil)
    ctx.requestRender()
  }

  public func onPointerUp(_ e: PointerInfo, _ ctx: ToolContext) { painting = false }

  public func cancel() {
    painting = false
    ghost = []
  }

  public func preview() -> ToolPreview {
    var p = ToolPreview()
    p.pendingPedestrians = ghost
    return p
  }
}
